Adding V1 spec draft: Scheduler(.postTask), TaskController, TaskSignal (#25)
WICG/scheduling-apis: spec/scheduling-tasks.md

   this priority will be used to schedule the task, and the task's priority is
   immutable. If left unspecified, the {{SchedulerPostTaskOptions/signal}}
   option will determine the task's priority if set, otherwise the priority
   defaults to {{TaskPriority/user-visible}}.

 - {{SchedulerPostTaskOptions/signal}}: An optional {{AbortSignal}} or
   {{TaskSignal}}. If a {{TaskSignal}} is specified and the
   {{SchedulerPostTaskOptions/priority}} is ommitted, the signal is used to
   determine the priority, which can be modified by the associated
   {{TaskController}}. The {{SchedulerPostTaskOptions/signal}} is also used to
   cancel pending tasks. See [Controlling Tasks](#sec-controlling-tasks) for
   details on using a {{TaskController}} or {{AbortController}} with
   {{Scheduler/postTask()}}.

 - {{SchedulerPostTaskOptions/delay}}: An optional delay (milliseconds) may be
   specified in order to delay the execution of the task.

{{Scheduler/postTask()}} returns a Promise that is fulfilled with the result of
the callback, or rejected with an "{{AbortError!!exception}}" {{DOMException}}
if the task is aborted.

<xmp class='idl'>
  dictionary SchedulerPostTaskOptions {

>       AbortSignal? signal = null;

Reviewer: Don't make dictionary members nullable. Instead just leave them as their default of non-required. This has slight observable effects in that `{ signal: null }` or `{ priority: null }` will throw, which is more idiomatic.

      TaskPriority? priority = null;
      long delay = 0;
  };

>   callback SchedulerPostTaskCallback = any ();

Reviewer: Yeah, I'm not sure. We can leave this as-is, although I'd keep an eye out for any other instances as this seems pretty generic and maybe it should be factored in to Web IDL at some point...

>   callback SchedulerPostTaskCallback = any ();

Author: Sounds good, and agreed (I was surprised there wasn't a built-in option for this).

[=scheduler task queues=] in response to a {{TaskSignal}}'s
<a for=TaskSignal>priority</a> changing, inserting based on
[=scheduler task/enqueue order=]. This approach would simplify
[=selecting the task queue of the next scheduler task=], but make priority
changes more complex.

A {{Scheduler}} object has a numeric <dfn for="Scheduler">next enqueue
order</dfn> which is initialized to 1.

Note: The [=Scheduler/next enqueue order=] is a strictly increasing number that
is used to determine task execution order across [=scheduler task queues=] of the
same {{TaskPriority}} within the same {{Scheduler}}. A logically equivalent
alternative would be to place the [=Scheduler/next enqueue order=] on the
[=event loop=], since the only requirements are that the number be strictly
increasing and not be repeated within a {{Scheduler}}.

Issue: Would it be simpler to just use a timestamp here?

The <dfn method for=Scheduler title="postTask(callback, options)">postTask(|callback|, |options|)</dfn>
method must return the result of [=scheduling a postTask task=] for [=this=]
given |callback| and |options|.

Definitions {#sec-scheduling-tasks-definitions}
---------------------

> ### Scheduler Tasks ### {#sec-def-scheduler-tasks}

Reviewer: Yeah, that seems reasonable to me.

> ### Scheduler Tasks ### {#sec-def-scheduler-tasks}

Author: Done.

> method must return the result of [=scheduling a postTask task=] for [=this=]

Reviewer: ```suggestion
method steps are to return the result of [=scheduling a postTask task=] for [=this=]
```

Author: Done.

  1. Set |task|'s <a attribute for="task">source</a> to |source|.
  1. Set |task|'s <a attribute for="task">document</a> to |document|.
  1. Set |task|'s <a attribute for="task">script evaluation environment settings
     object set</a> to a new empty [=set=].
  1. [=set/Append=] |task| to |queue|'s [=scheduler task queue/tasks=].
  1. Return |task|.

  Issue: We should consider refactoring the HTML spec to add a constructor for
  <a for="/">task</a>. One problem is we need the new task to be a
  [=scheduler task=] rather than a <a for="/">task</a>.
</div>

<div algorithm>
  To <dfn for="scheduler task queue">remove</dfn> a [=scheduler task=] |task| from
  [=scheduler task queue=] |queue|, [=set/remove=] |task| from |queue|'s
  [=scheduler task queue/tasks=].
</div>

### Scheduling Tasks ### {#sec-scheduler-alg-scheduling-tasks}

<div algorithm="schedule a postTask task">
  To <dfn lt="schedule a postTask task|scheduling a postTask task">schedule a postTask task</dfn>
  for {{Scheduler}} |scheduler| given a {{SchedulerPostTaskCallback}} |callback| and
  {{SchedulerPostTaskOptions}} options, run the following steps:

  1. Let |result| be [=a new promise=].

>   1. Let |signal| be |options|["{{SchedulerPostTaskOptions/signal}}"].

Reviewer: If you follow the above advice about not using nullable options, then you'll need to change these a bit to check for whether the value `[=map/exists=]`. You can convert not-present into null here if you want, so that it's easier to pass it through the algorithms.

       that started before this one, and whose |delay| is equal to or less
       than this one's, have completed.
    1. Optionally, wait a further [=implementation-defined=] length of time.
    1. [=Schedule a task to invoke a callback=] for |scheduler| given |queue|,
       |signal|, |callback|, and |result|.
  1. Otherwise the task is not delayed. [=Schedule a task to invoke a callback=]
     for |scheduler| given |queue|, |signal|, |callback|, and |result|.
</div>

Issue: We need to figure out exactly how we want to spec delayed tasks, and if
we can refactor the timer spec to use a common method. As written, this uses
steps 15&ndash;17 of the timer initialization steps algorithm, but there are a
couple things we might want to change: (1) how to account for suspend? (2) how
to account for current throttling techniques (see also
[this issue](https://github.com/whatwg/html/issues/5925))?

<div algorithm="select the scheduler task queue">
  To <dfn lt="select the scheduler task queue|selecting the scheduler task queue">select the scheduler task queue</dfn>
  for a {{Scheduler}} |scheduler| given ({{TaskSignal}} or {{AbortSignal}})
  |signal|, and a {{TaskPriority}} |priority|:

>   1. If |priority| is null and |signal| is not null and |signal| is a

Reviewer: You can be extra-precise by using `|signal| [=implements=] the {{TaskSignal}} interface`

     {{TaskSignal}} object, then
    1. If |scheduler|'s [=Scheduler/dynamic priority task queue map=] does not
       [=map/contain=] |signal|, then
      1. Let |queue| be the result of [=creating a scheduler task queue=] given
         |signal|'s <a for=TaskSignal>priority</a>.
      1. Set [=Scheduler/dynamic priority task queue map=][|signal|] to |queue|.
      1. <a for=TaskSignal>Add a priority change algorithm</a> to |signal| that
         runs the following steps:
        1. Set |queue|'s [=scheduler task queue/priority=] to |signal|'s
           {{TaskSignal/priority}}.
    1. Return [=Scheduler/dynamic priority task queue map=][|signal|].
  1. Otherwise |priority| is used to determine the task queue:

>     1. If |priority| is null, set |priority| to {{TaskPriority/user-visible}}.

Reviewer: ```suggestion
    1. If |priority| is null, set |priority| to "{{TaskPriority/user-visible}}".
```

Stylistically I prefer making it clear that enum values are strings in this way.

    1. If |scheduler|'s [=Scheduler/static priority task queue map=] does not
       [=map/contain=] |priority|, then
      1. Let |queue| be the result of [=creating a scheduler task queue=] given
         |priority|.
      1. Set [=Scheduler/static priority task queue map=][|priority|] to |queue|.
    1. Return [=Scheduler/static priority task queue map=][|priority|].
</div>

<div algorithm>
  To <dfn>schedule a task to invoke a callback</dfn> for {{Scheduler}}

>   |scheduler| given a [=scheduler task queue=] |queue|, an {{AbortSignal}} |signal|,

Reviewer: ```suggestion
  |scheduler| given a [=scheduler task queue=] |queue|, an {{AbortSignal}} or null |signal|,
```

>   SchedulerPostTaskCallback |callback|, and promise |result|:

Reviewer: ```suggestion
  a {{SchedulerPostTaskCallback}} |callback|, and a promise |result|:
```

Reviewer: Hmm yeah, historically I have mostly talked about the "conceptual" promise, but I guess it is also a formal type name `Promise`. So capitalizing and linking is probably a bit better.
